File: RemoveDuplicatesfromSortedArray.java
/*
Runtime: 1 ms, faster than 100.00% of Java online submissions for Remove Duplicates from Sorted Array.
Memory Usage: 47.7 MB, less than 62.68% of Java online submissions for Remove Duplicates from Sorted Array.
*/
// Optimised Solution: TC: O(N) SC: O(1)
class Solution {
    public int removeDuplicates(int[] nums) {
        
        int index = 1;
        
        for(int i=1; i<nums.length; i++){
            if(nums[i] != nums[i-1]){
                nums[index++] = nums[i];
            }
        }
        
        return index;
        
    }
}

File: RemoveDuplicatesfromSortedArray.js
// Optimized Solution TC: O(N) SC: O(1)
/**
 * @param {number[]} nums
 * @return {number}
 */
var removeDuplicates = function(nums) {
    let i = 0;
    nums.forEach(num => {
        if(i === 0 || num !== nums[i - 1]) {
            nums[i] = num;
            i++; 
        }
    });
    return i;
};


/**
 * @param {number[]} nums
 * @return {number}
 */
var removeDuplicates = function(nums) {
    let index = 0, i = 0;
    while(i < nums.length) {
        nums[index++] = nums[i];
        while(nums[i] === nums[i+1]) i++;
        i++;
    }
    return index;
};

// Another Solution
/**
 * @param {number[]} nums
 * @return {number}
 */
var removeDuplicates = function(nums) {
    let index = 1;
    let currentElement = nums[0];
    nums.forEach((el, i) => {
        if(el !== currentElement) {
            currentElement = el;
            nums[index] = el;
            index++;
        }
    });
    return index;
};
